#### Author : Mithra Chintha   And  Date : 3/7/2017

In [1]:
#Importing Pandas, numpy and matplotlib packages
import pandas as pd
import numpy as np
from pandas import DataFrame,Series
import matplotlib.pyplot as mpl
%matplotlib inline
from __future__ import division

In [175]:
ls

 Volume in drive C is Windows
 Volume Serial Number is 9685-0F91

 Directory of C:\Users\chint\Documents\Python\Indeed

03/07/2017  21:02    <DIR>          .
03/07/2017  21:02    <DIR>          ..
03/05/2017  12:45    <DIR>          .ipynb_checkpoints
03/07/2017  16:43         3,596,178 AgeInjuriesboxplot.html
01/17/2016  12:38               347 BodyParts.csv
03/07/2017  16:53            37,354 DataAssesmentIndeed.ipynb
01/17/2016  12:19               759 DiagnosisCodes.csv
01/17/2016  12:44               377 Disposition.csv
01/16/2016  18:36        10,098,911 NEISS2014.csv
03/07/2017  21:01         3,791,201 Q5DataAnalysis.twbx
03/07/2017  21:02           870,130 Report.docx
               8 File(s)     18,395,257 bytes
               3 Dir(s)  311,073,361,920 bytes free


### Import Data Sets provided

In [177]:
#Import the NEISS2014 csv data file
neiss2014=pd.read_csv("NEISS2014.csv")

In [178]:
#To Check Null values in the neiss2014 data set
neiss2014.isnull().sum()

CPSC Case #        0
trmt_date          0
psu                0
weight             0
stratum            0
age                0
sex                0
race               0
race_other     61922
diag               0
diag_other     57380
body_part          0
disposition        0
location           0
fmv                0
prod1              0
prod2          57053
narrative          0
dtype: int64

#### Body parts and Diagnosis information is provided in codes and the actual values of each of these codes are provided in seperate datasets

In [4]:
#Import the Body parts data file
bodyParts=pd.read_csv("BodyParts.csv")

In [6]:
 #Import Diagnosis Code data file
diagnosisCode=pd.read_csv("DiagnosisCodes.csv")

In [8]:
#Check the data in neiss2014
neiss2014.iloc[0:10,]

CPSC Case # trmt_date  psu    weight stratum  age     sex         race  \
0    141200216  10/23/14   63   99.7040       M   21  Female  None listed   
1    140117851    1/4/14   63   81.5760       M   62  Female        White   
2    150144993   8/17/14   63   99.7040       M   21  Female  None listed   
3    150230176   7/27/14   63   99.7040       M   30  Female  None listed   
4    141220717    9/8/14   63   99.7040       M   16    Male  None listed   
5    150151229   8/12/14   63   99.7040       M   22    Male  None listed   
6    150216968   7/15/14   63   99.7040       M   92    Male  None listed   
7    141200989  11/24/14   63  112.1671       M   89    Male  None listed   
8    141241629  12/11/14   63   99.7040       M   75  Female  None listed   
9    150226548   7/14/14   63   99.7040       M   68    Male  None listed   

  race_other  diag diag_other  body_part  disposition  location  fmv  prod1  \
0        NaN    62        NaN         75            1         1    0    679   
1        NaN    57        NaN         79            1         1    0   1807   
2        NaN    57        NaN         83            1         0    0   1333   
3        NaN    64        NaN         79            1         1    0   4076   
4        NaN    57        NaN         82            1         0    0   1893   
5        NaN    64        NaN         79            1         1    0   1871   
6        NaN    72        NaN         36            1         1    0    214   
7        NaN    57        NaN         79            4         1    0   4074   
8        NaN    59        NaN         75            1         1    0   3251   
9        NaN    64        NaN         35            1         4    0   5040   

    prod2                                          narrative  
0  1807.0  21YOF LAYING ON THE COUCH AND ROLLED OFF TO TH...  
1   115.0  62YOF TRIPPED OVER THE CORD FOR THE VACUUM CLE...  
2     NaN    21YOF FELL OFF A SKATEBOARD AND FRACTURED FOOT   
3     NaN  30YOF TRYING TO GET OUT OF BED AND FELT A POP ...  
4     NaN   16YOM HAD HAND SLAMMED IN A DOOR FRACTURED HAND   
5     NaN  22YOM HELPING FATHER PUT UP A FENCE AND STRAIN...  
6     NaN  92YOM BUMPED LOWER LEG ON A DISHWASHER DOOR SU...  
7     NaN  89YOM GETTING OUT OF A CHAIR AND FELL ONTO HIP...  
8     NaN  75YOF GETTING INTO A BUILT IN GROUND SWIMMING ...  
9     NaN        68YOM RIDING HIS BICYCLE AND STRAINED KNEE

In [9]:
#Merge the diagnosis Code in the neiss2014
neiss2014 = pd.merge(neiss2014,diagnosisCode,how='left',left_on='diag',right_on='Code').drop(['Code'],axis=1)

In [10]:
#Merge the Bodyparts in the neiss2014
neiss2014 = pd.merge(neiss2014,bodyParts,how='left',left_on='body_part',right_on='Code').drop(['Code'],axis=1)

In [179]:
#Check null values after merging the data sets
neiss2014.isnull().sum()

CPSC Case #        0
trmt_date          0
psu                0
weight             0
stratum            0
age                0
sex                0
race               0
race_other     61922
diag               0
diag_other     57380
body_part          0
disposition        0
location           0
fmv                0
prod1              0
prod2          57053
narrative          0
dtype: int64

# Question 1

#### What are the top three body parts most frequently represented in this dataset? What are the top three body parts that are least frequently represented?

#### Approach: What are the top three body parts most frequently represented in this dataset?

value_counts() is a function in Pandas Series which gives the counts of unique values. By default, this function gives the descending order. BodyPart column is selected from the dataframe neiss2014. Used the head() function with 3 as input returns top 3 in the list i.e., the top three most frequent body parts 

In [25]:
neiss2014.BodyPart.value_counts().head(3)

Head      9891
Face      5786
Finger    5783
Name: BodyPart, dtype: int64

The top three body parts are 
Head,
Face,
Finger

#### Approach: What are the top three body parts that are least frequently represented?

Similar to the head() function, to get the top three least frequently represent tail() function can be used.

In [24]:
neiss2014.BodyPart.value_counts().tail(3)

Not Recorded      390
Pubic region      286
25-50% of body      4
Name: BodyPart, dtype: int64

From the above results, top three has 'Not Recorded'. So to get the top three body parts the 4th one has to be selected

In [26]:
neiss2014.BodyPart.value_counts().tail(4)

Internal          549
Not Recorded      390
Pubic region      286
25-50% of body      4
Name: BodyPart, dtype: int64

The top three body parts are 
Internal, 
Public region,
25- 50% of body

# Question 2

#### How many injuries in this dataset involve a skateboard?Of those injuries, what percentage were male and what percentage were female? What was the average age of someone injured in an incident involving a skateboard?


#### Approach: How many injuries in this dataset involve a skateboard

Used a variable called narrative which contains information about how an injury occured. To check the injuries involved a skateboard, I used this variable to search for occurences of 'skateboard' or 'skate board'. In the below code, narrative column is selected from the data frame neiss2014 and is type casted to string object. This is passed to lower() function to be make sure that all the occurences in the data frame are in lower case. This again be type casted to string and is passed to contains() function to check the occurances of the string 'skateboard' or 'skate board'. This will result in True/False output to denote whether each record has the string skateboard/skate board or not. This result is passed to sum() function to count the number of occurances of skateboard/skate board. During this computation True will be encoded as 1 and False will be encoded as 0.

In [38]:
neiss2014.narrative.str.lower().str.contains("skate board|skateboard").sum()

472

Therefore there are 472 injuries involved in skateboard

#### Approach: Of those injuries, what percentage were male and what percentage were female?

To find out the percentages of male and female injuries out of all the skateboard injuries, sex variable is selected and passed to value_counts(normalize=True) function. if we pass normalize=True parameter, the function returns proportions and not the actual counts.

In [43]:
neiss2014[neiss2014.narrative.str.lower().str.contains("skate board|skateboard")].sex.value_counts(normalize=True)*100

Male      82.627119
Female    17.372881
Name: sex, dtype: float64

Therefore, 82.6% of the skateboard injuries are Males, 17.37% of the skateboard injuries are Females

#### Approach: What was the average age of someone injured in an incident involving a skateboard?

To find the average age of someone with skateboard injuries, age variable is selected for all the cases with skateboard injuries and is passed to the mean() function

In [44]:
neiss2014[neiss2014.narrative.str.lower().str.contains("skate board|skateboard")].age.mean()

17.953389830508474

Therefore, average age is near to 18 years

# Question 3

#### What diagnosis had the highest hospitalization rate? What diagnosis most often concluded with the individual leaving without being seen? Briefly discuss your findings and any caveats you'd mention when discussing this data

#### Approach: What diagnosis had the highest hospitalization rate?

Proportion of various disposition categories like Treated, Transferred to other hospital, Hospitilized are calculated for each diagnosis and the percentages of Hospitilized cases(coded as 4) per each diagnosis is selected. The results are finally sorted in the descending order inorder to find out the diagnosis with he highest hospitilization rate and displayed the highest hospitalization rate

In [99]:
pd.DataFrame({'DispositionRate':neiss2014[['Diagnosis','disposition']].groupby(['Diagnosis'])['disposition']\
          .apply(lambda x: pd.value_counts(x)/x.count().astype(float))}).\
xs(4,level=1).sort_values(['DispositionRate'],ascending=False).head(1)

DispositionRate
Diagnosis                                       
Submersion (including Drowning)         0.425926

As shown above, submersion has the highest hospitalization rate

#### Approach: What diagnosis most often concluded with the individual leaving without being seen?

In the code below, I have calculated the counts of individuals leaving without being seen per each diagnosis. Pivot table functionality in Pandas is used to solve this question and Diagnosis and Disposition variables are selected as index values. Aggregation funciton "count" is selected to calculated the counts. Since we are only interested in the counts of cases leaving without being seen, xs(6,level=1) function is used to fetch only those cases.
As shown below, Laceration has most individuals leaving without being seen.

In [110]:
pd.DataFrame(neiss2014.pivot_table(index=['Diagnosis','disposition'],aggfunc='count')['diag']).\
xs(6,level=1).sort_values(['diag'],ascending=False).head(2)

diag
Diagnosis             
Other/Not Stated   257
Laceration         123

As shown above, Laceration diagnosis most often concluded with individuals leaving without being seen 

#### Approach: Briefly discuss your findings and any caveats you'd mention when discussing this data

Apart from the counts of individuals, percentages are also calculated below. Proportion of various disposition categories like Treated, Transferred to other hospital, Hospitalized are calculated for each diagnosis and the percentages of Cases without being seen(coded at 6) per each diagnosis is selected. The results are finally sorted in the descending order inorder to find out the diagnosus most often concluded with the individual leaving without being seen.


In [114]:
pd.DataFrame({'DispositionRate':neiss2014[['Diagnosis','disposition']].groupby(['Diagnosis'])['disposition']\
          .apply(lambda x: pd.value_counts(x)/x.count().astype(float))}).\
xs(6,level=1).sort_values(['DispositionRate'],ascending=False).head(3)

DispositionRate
Diagnosis                                
Poisoning                        0.033149
Other/Not Stated                 0.031650
Aspirated foreign object         0.030303

As shown Above, Poisoning has the highest rate of individuals leaving without being seen.

# Question 4

#### Approach: Visualize any existing relationship between age and reported injuries

In [145]:
sorted(neiss2014.age.unique(),reverse=True)

[223,
 222,
 221,
 220,
 219,
 218,
 217,
 216,
 215,
 214,
 213,
 212,
 211,
 210,
 209,
 208,
 207,
 206,
 205,
 204,
 203,
 202,
 201,
 105,
 103,
 102,
 101,
 100,
 99,
 98,
 97,
 96,
 95,
 94,
 93,
 92,
 91,
 90,
 89,
 88,
 87,
 86,
 85,
 84,
 83,
 82,
 81,
 80,
 79,
 78,
 77,
 76,
 75,
 74,
 73,
 72,
 71,
 70,
 69,
 68,
 67,
 66,
 65,
 64,
 63,
 62,
 61,
 60,
 59,
 58,
 57,
 56,
 55,
 54,
 53,
 52,
 51,
 50,
 49,
 48,
 47,
 46,
 45,
 44,
 43,
 42,
 41,
 40,
 39,
 38,
 37,
 36,
 35,
 34,
 33,
 32,
 31,
 30,
 29,
 28,
 27,
 26,
 25,
 24,
 23,
 22,
 21,
 20,
 19,
 18,
 17,
 16,
 15,
 14,
 13,
 12,
 11,
 10,
 9,
 8,
 7,
 6,
 5,
 4,
 3,
 2,
 0]

Age variable is measured in years for cases between 2 and 115 years. For the cases <2 years, the data is encoded in months with an additional code beginning with 2. Therefore, for cases <2 years, the data will have three digits. Inorder to bring consistancy in the data, all the data is converted into months in the below code.
In the below code, age will be multiplied with 12 if <115. Else remainder of age with 200 is calculated. Remainder of age with 200 will return the number of months.

In [147]:
neiss2014.age=neiss2014.age.map(lambda x: x*12 if x<115 else x%200)

For readability of x-axis labels in the graph, extra information in the diagnosis variable is discarded

In [150]:
neiss2014.Diagnosis=neiss2014.Diagnosis.str.replace(r"\(.*\)","")

Bokeh package is used to visualize the relationship between age and reported injuries. Import the important components in the bokeh package


In [169]:
from bokeh.charts import BoxPlot, output_file, show
from bokeh.charts import defaults, vplot, hplot

In [173]:
#default width and height of the graphs are et
defaults.width = 1400
defaults.height = 750

box_plot = BoxPlot(neiss2014, label='Diagnosis', values='age', color='blue',
                   title="Injury vs Age")



In [174]:
#Download the boxplot in html format with filename as ageinjuriesboxplot.html
output_file("AgeInjuriesboxplot.html")
show(box_plot)

INFO:bokeh.core.state:Session output file 'AgeInjuriesboxplot.html' already exists, will be overwritten.


# Question 5+

#### Investigate the data however you like and discuss any interesting insights you can find in the data

Added in the report